src/actix/certificate_helpers.rs

```diff
@@ -0,0 +1,138 @@
+use std::fs;
+use std::time::{Duration, Instant};
+
+use openssl::error::ErrorStack;
+use openssl::ssl::{
+    SniError, SslAcceptor, SslAcceptorBuilder, SslContext, SslContextBuilder, SslFiletype,
+    SslMethod, SslVerifyMode,
+};
+use openssl::x509::store::X509StoreBuilder;
+use openssl::x509::X509;
+use parking_lot::RwLock;
+
+use crate::settings::{Settings, TlsConfig};
+
+struct SslContextHolder {
+    ssl_context: SslContext,
+    tls_config: TlsConfig,
+    verify_client_cert: bool,
+    refresh_interval: Duration,
+    last_updated: Instant,
+}
+
+impl SslContextHolder {
+    fn new(
+        tls_config: &TlsConfig,
+        verify_client_cert: bool,
+        refresh_interval: Duration,
+    ) -> Result<SslContextHolder, ErrorStack> {
+        let ssl_context = build_ssl_context(tls_config, verify_client_cert)?;
+        Ok(SslContextHolder {
+            ssl_context,
+            tls_config: tls_config.clone(),
+            verify_client_cert,
+            refresh_interval,
+            last_updated: Instant::now(),
+        })
+    }
+
+    fn try_get_ssl_context(&self) -> Option<&SslContext> {
+        if self.last_updated.elapsed() < self.refresh_interval {
+            Some(&self.ssl_context)
+        } else {
+            None
+        }
+    }
+
+    fn refresh_and_get_ssl_context(&mut self) -> Result<&SslContext, ()> {
+        if self.last_updated.elapsed() < self.refresh_interval {
+            Ok(&self.ssl_context)
+        } else {
+            self.refresh()?;
+            Ok(&self.ssl_context)
+        }
+    }
+
+    fn refresh(&mut self) -> Result<(), ()> {
+        log::info!("Refreshing TLS certificates for actix!");
+        match build_ssl_context(&self.tls_config, self.verify_client_cert) {
+            Ok(ssl_context) => {
+                self.ssl_context = ssl_context;
+                Ok(())
+            }
+            Err(_err) => Err(()),
+        }
```

Comment:
This is a bit more concise:

```suggestion
        self.ssl_context =
            build_ssl_context(&self.tls_config, self.verify_client_cert).map_err(|_| ())?;
        Ok(())
```